[Detekt Baseline Warnings] FluxC Module - Resolve/Suppress Complexity & Naming Warnings

ParaskP7 opened the pull request:
Parent #2479
Partially Closes: #2485

This PR resolves/suppresses all `complexity` and `naming` related warnings for the `fluxc` module:

- `59` x complexity
    1. `6` x ComplexCondition (Resolve: 6bee094fce82a9395e81e3db309cba44f7b29f95 + Suppress: 11308f56b2b4de608c84a568c94253c5f5a9cc8d)
    2. `14` x ComplexMethod (Resolve: c13914cbf9aa742bd51849fcdbbe237b829fcdfe + Suppress: e7556975ebcccd755f06a2d49e3610b3b6909811 + 3c13e594043c51cbb0f4dbeda5fefbc8d2434d11)
    3. `4` x LargeClass (Suppress: 13743875195599f107e18067210122a3eb2452ae)
    4. `6` x LongMethod (Suppress: cace7134f5a853c5be73c765a9cdf611146bc6c9)
    5. `21` x LongParameterList (Suppress: 96876e2aa1e7525b0e482238feb152ed1c910913)
    6. `8` x NestedBlockDepth (Resolve: bc4768c57b57ef0678c0f6ee4bc81c1771133c28 + Suppress: bc67aa2e0e4c16e1c3fe348916557f2062f06e7f)
- `51` x naming
    1. `31` x ConstructorParameterNaming (Suppress: a7553aa6ca12e966cf12eba56e019c5f069354ac)
    2. `20` x VariableNaming (Resolve: 8ef999cb2da951f270b293ac276a5ef41977b9b7 + Suppress: 8576358288f33279102620184e709f56f1030d3f)

PS.1: I am randomly adding one engineer from the WPAndroid team and one from the WCAndroid team as the main reviewers, that is, in addition to the @wordpress-mobile/owl-team team itself, since I just want someone from these teams to ~~sing-off (😅)~~ sign-off on that change as well (or additionally to the 🦉 team).

PS.2: I recommend reviewing this PR commit-by-commit.

-----

To test:

- There is nothing much to test here.
- Verifying that all the CI checks are successful should be enough (especially the `detekt` check).
- However, if you really want to be thorough, you could smoke test the `example` and/or the `WPAndroid/WCAndroid` apps to verify that everything works as expected.

PS: Please ignore the failing `connected-tests` check as they are unrelated to this PR. This check has been failing consistency for some time now due to tests being flaky or needing to be updated/fixed.

ParaskP7 committed:
Analysis: Resolve complex condition warn for latest post insights store

Description: "Complex conditions make it hard to understand which cases
lead to the condition being true or false. To improve readability and
understanding of complex conditions consider extracting them into
well-named functions or variables and call those instead."

For more info see: https://detekt.dev/docs/rules/complexity
#complexcondition

ParaskP7 committed:
Analysis: Suppress complex condition warn for fluxc module

Description: "Complex conditions make it hard to understand which cases
lead to the condition being true or false. To improve readability and
understanding of complex conditions consider extracting them into
well-named functions or variables and call those instead."

For more info see: https://detekt.dev/docs/rules/complexity
#complexcondition

ParaskP7 committed:
Analysis: Resolve complex method warning for scan sql utils

Description: "Complex methods are hard to understand and read. It might
not be obvious what side-effects a complex method has. Prefer splitting
up complex methods into smaller methods that are in turn easier to
understand. Smaller methods can also be named much clearer which leads
to improved readability of the code.

This rule uses McCabe's Cyclomatic Complexity (MCC) metric to measure
the number of linearly independent paths through a function's source
code (https://www.ndepend.com/docs/code-metrics#CC). The higher the
number of independent paths, the more complex a method is. Complex
methods use too many of the following statements. Each one of them adds
one to the complexity count."

For more info see: https://detekt.dev/docs/rules/complexity
#complexmethod

ParaskP7 committed:
Analysis: Suppress complex method warning for fluxc module

Description: "Complex methods are hard to understand and read. It might
not be obvious what side-effects a complex method has. Prefer splitting
up complex methods into smaller methods that are in turn easier to
understand. Smaller methods can also be named much clearer which leads
to improved readability of the code.

This rule uses McCabe's Cyclomatic Complexity (MCC) metric to measure
the number of linearly independent paths through a function's source
code (https://www.ndepend.com/docs/code-metrics#CC). The higher the
number of independent paths, the more complex a method is. Complex
methods use too many of the following statements. Each one of them adds
one to the complexity count."

For more info see: https://detekt.dev/docs/rules/complexity
#complexmethod

ParaskP7 committed:
Analysis: Suppress long method warning for insights mapper

Description: "Methods should have one responsibility. Long methods can
indicate that a method handles too many cases at once. Prefer smaller
methods with clear names that describe their functionality clearly.

Extract parts of the functionality of long methods into separate,
smaller methods."

For more info see: https://detekt.dev/docs/rules/complexity#longmethod

ParaskP7 committed:
Analysis: Resolve checkstyle store warning for plugin coroutine store

Warning Message: "Stores must be annotated with @Singleton"

For more info see: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/blob/trunk/config/checkstyle.xml#L84-L88

ParaskP7 committed:
Analysis: Suppress large class warnings for fluxc module

Description: "This rule reports large classes. Classes should generally
have one responsibility. Large classes can indicate that the class does
instead handle multiple responsibilities. Instead of doing many things
at once prefer to split up large classes into smaller classes. These
smaller classes are then easier to understand and handle less things."

For more info see: https://detekt.dev/docs/rules/complexity#largeclass

ParaskP7 committed:
Analysis: Suppress long parameter list warnings for fluxc module

Description: "Reports functions and constructors which have more
parameters than a certain threshold."

For more info see: https://detekt.dev/docs/rules/complexity
#longparameterlist

ParaskP7 committed:
Analysis: Resolve nested block depth warnings for fluxc module

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Analysis: Suppress nested block depth warnings for fluxc module

Description: "This rule reports excessive nesting depth in functions.
Excessively nested code becomes harder to read and increases its hidden
complexity. It might become harder to understand edge-cases of the
function.

Prefer extracting the nested code into well-named functions to make it
easier to understand."

For more info see: https://detekt.dev/docs/rules/complexity
#nestedblockdepth

ParaskP7 committed:
Analysis: Suppress constructor parameter naming warnings for fluxc mod.

Description: "Reports constructor parameter names that do not follow the
specified naming convention."

For more info see: https://detekt.dev/docs/rules/naming
#constructorparameternaming

ParaskP7 committed:
Analysis: Resolve variable naming warning for react native store

Description: "Reports variable names that do not follow the specified
naming convention."

For more info see: https://detekt.dev/docs/rules/naming#variablenaming

ParaskP7 committed:
Analysis: Resolve variable naming warning for fluxc module

Description: "Reports variable names that do not follow the specified
naming convention."

For more info see: https://detekt.dev/docs/rules/naming#variablenaming

irfano commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/PluginCoroutineStore.kt:
```diff
@@ -29,11 +29,12 @@ import org.wordpress.android.fluxc.utils.CurrentTimeProvider
 private const val PLUGIN_CONFIGURATION_DELAY = 1000L
 
-class PluginCoroutineStore
-@Inject constructor(
+@Singleton
+class PluginCoroutineStore @Inject constructor(
```
This is my only concern. Adding `@Singleton` will change the function though it's the correct change. 
But I couldn't find any problem about that.

ParaskP7 replied:
👋 @irfano and thank you so much for actually catching that and opening a discussion about it! 🙇 

I also was on the fence about this change, but I chose to progress with it anyway since this is the rule. Yes, it will change the function in terms of it will now be just one instance of this store instead of multiple of them, but this might very well not change the outcome, but then again it might do, though I doubt it betting on the rule, which explicitly states that all stores should be singletons and as such clients should depend on that to be the case.

PS: I am now shamelessly pinging @planarvoid on that since he was the initial and only author of that store. It's been a while since then, a bit more than a year, but he might know if this was done deliberately, that is, to not have this store as singleton.

ParaskP7 replied:
The interesting bit about [this CheckStyle configuration and thus regex](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/config/checkstyle.xml#L84-L88) is the fact that it is **NOT** correctly working when the store `class` and `constructor` are split into two separate lines, like that:

```
class PluginCoroutineStore
@Inject constructor(
    ...
) {
    ...
}
```

It only works if the `class` and `constructor` are on the same line, like this:

```
class PluginCoroutineStore @Inject constructor(
    ...
) {
    ...
}
```

Thus, and because I did this refactor via [this commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/e7556975ebcccd755f06a2d49e3610b3b6909811#diff-44f37195454337f4b258b698820c5b5fa052a6ee07cc9d21d743e479868b4784), CheckStyle started complaining, triggering me to apply [this additional change](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2491/commits/3c13e594043c51cbb0f4dbeda5fefbc8d2434d11) we are now discussing.

I wonder how many other such store related CheckStyle warnings are currently not being reported because of the `class` and `constructor` being split into two separate lines. 🤔

ParaskP7 replied:
Interestingly enough, I did a global search on `class .*Store` and then checked each store one-by-one to see if we have other such stores that are not defined as singletons. The results is that in addition to [PluginCoroutineStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e5cf7a211c05ac9c2051415b5221924918a0b589/fluxc/src/main/java/org/wordpress/android/fluxc/store/PluginCoroutineStore.kt#L35-L36), we have only 2 other such stores (3 in total, out of hundreds of them), the [SiteOptionsStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e5cf7a211c05ac9c2051415b5221924918a0b589/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteOptionsStore.kt#L36-L37) and [XPostsStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e2752ac62e012e5ad38ee3ae36acfc506d01ea0b/fluxc/src/main/java/org/wordpress/android/fluxc/store/XPostsStore.kt#L14-L15), and all 3 of them are having their `class` and `constructor` defined into two separate lines. 🤔 

The history on all those 3 stores shows that when those stores were created (initial commit) they were created **WITHOUT** the `@Singleton` annotation. To me this smells that this wasn't intentional, but rather an oversight, especially since CheckStyle wasn't complaining about that (as it should have had). 🤔 x 10

Wdyt?

Cc @irfano @planarvoid

ParaskP7 replied:
Also, from @planarvoid [PR's description here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2043) I don't see any mention about `PluginCoroutineStore` and it not being a singleton for some particular reason. Same applies to the other 2 such stores:

- [Add site homepage settings #1583](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1583) for the [SiteOptionsStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e5cf7a211c05ac9c2051415b5221924918a0b589/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteOptionsStore.kt#L36-L37)
- [Add XPostsStore for Retrieving a site's xpost candidates #1769](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1769) for the [XPostsStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e2752ac62e012e5ad38ee3ae36acfc506d01ea0b/fluxc/src/main/java/org/wordpress/android/fluxc/store/XPostsStore.kt#L14-L15)

irfano replied:
Good findings!
>The interesting bit about [this CheckStyle configuration and thus regex](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/config/checkstyle.xml#L84-L88) is the fact that it is NOT correctly working

Firstly, we can try to fix the regex. Maybe we can open an issue for now, and fix it in another PR.
>The results is that in addition to [PluginCoroutineStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e5cf7a211c05ac9c2051415b5221924918a0b589/fluxc/src/main/java/org/wordpress/android/fluxc/store/PluginCoroutineStore.kt#L35-L36), we have only 2 other such stores (3 in total, out of hundreds of them), the [SiteOptionsStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e5cf7a211c05ac9c2051415b5221924918a0b589/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteOptionsStore.kt#L36-L37) and [XPostsStore](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/e2752ac62e012e5ad38ee3ae36acfc506d01ea0b/fluxc/src/main/java/org/wordpress/android/fluxc/store/XPostsStore.kt#L14-L15)

I'm trying to understand the use of stores and thinking about why we need to make stores singleton. They don't have any variables that must be the same for consumers. They only have parameters coming from the constructor. And I guess the critical parameters are [`XPostsSqlUtils`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/XPostsSqlUtils.kt) and [`Dispatcher`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/Dispatcher.java) because these are writing and emitting results. So, to my understanding, `XPostsSqlUtils` and `Dispatcher` need to be a singleton, and they are already singleton, even in different instances of `PluginCoroutineStore`, for example. I feel like `@Singleton` could be unnecessary, but I haven't tested it. I am just telling my thoughts.
I don't have much experience on this repo, but I'd like to know why the `@Singleton` rule was added in checkstyle.

ParaskP7 replied:
I think for us to unblock this PR but still have this discussion going, I am going to revert this `@Singleton` change for now, merge this PR and open an issue about those 3 stores (them not being singletons) so we can continue the discussion there. Wdyt? 🤔 

> Firstly, we can try to fix the regex. Maybe we can open an issue for now, and fix it in another PR.

I agree, I'll create an additional ticket for that too. 💯 

> I'm trying to understand the use of stores and thinking about why we need to make stores singleton. They don't have any variables that must be the same for consumers. They only have parameters coming from the constructor. And I guess the critical parameters are [XPostsSqlUtils](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/persistence/XPostsSqlUtils.kt) and [Dispatcher](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/fluxc/src/main/java/org/wordpress/android/fluxc/Dispatcher.java) because these are writing and emitting results. So, to my understanding, XPostsSqlUtils and Dispatcher need to be a singleton, and they are already singleton, even in different instances of PluginCoroutineStore, for example. I feel like @Singleton could be unnecessary, but I haven't tested it. I am just telling my thoughts.
I don't have much experience on this repo, but I'd like to know why the @Singleton rule was added in checkstyle.

Yes, I agree, it seems a bit strange indeed. Maybe we can search on P2s or Slack about that, just to figure out the real reasoning behind that, I am sure there was a good reason... 🤔

ParaskP7 replied:
👋 @irfano !

> I think for us to unblock this PR but still have this discussion going, I am going to revert this @Singleton change for now, merge this PR...

PS: I reverted this change for now (see 6015d6bdcfb895a6fbb0ddb6b33dff9e677fd088) and will merge this PR as soon as CI completes successfully. ✅

ParaskP7 replied:
> ...and open an issue about those 3 stores (them not being singletons) so we can continue the discussion there.

Also, I just opened [this issue](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2495) and assigned all us 3 to it for the time being.

ParaskP7 replied:
> I agree, I'll create an additional ticket for that too.

As discussed, I just opened [this issue](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2497) too and assigned us 2 to it (plus @aforcier, the initial and only author of this [rule and regex](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/config/checkstyle.xml#L84-L88).

irfano replied:
>I think for us to unblock this PR but still have this discussion going, I am going to revert this @Singleton change for now, merge this PR and open an issue about those 3 stores (them not being singletons) so we can continue the discussion there.

Good call.

Thanks for opening new issues.

irfano reviewed (approved):
Well-defined commits. Great job, @ParaskP7!
LGTM! 👍🏻 I'm approving, but UI tests are failing. I'm not sure we can temporarily ignore UI tests like in "WPAndroid".

ParaskP7 commented:
👋 @irfano !

Thank you so much for reviewing and testing this PR! 🥇 🙇 🌟 

> LGTM! 👍🏻 I'm approving, but UI tests are failing. I'm not sure we can temporarily ignore UI tests like in "WPAndroid".

Yes, we can actually safely ignore these connected tests for now (see Slack message: `C02QANACA/p1659504296491789`) and that's why I added this `PS` about them in this PR's description (see below).

```
PS: Please ignore the failing connected-tests check as they are unrelated to this PR. This check has
been failing consistency for some time now due to tests being flaky or needing to be updated/fixed.
```

ParaskP7 committed:
Analysis: Revert making plugin coroutine store a singleton

For more info follow discussion: https://github.com/wordpress-mobile/
WordPress-FluxC-Android/pull/2491#discussion_r943548751
